executor/src/witgen/evaluators/expression_evaluator.rs

```diff
@@ -0,0 +1,133 @@
+use std::collections::BTreeMap;
+
+use powdr_ast::analyzed::{
+    AlgebraicBinaryOperation, AlgebraicBinaryOperator, AlgebraicExpression as Expression,
+    AlgebraicReference, AlgebraicReferenceThin, AlgebraicUnaryOperation, AlgebraicUnaryOperator,
+    Analyzed, PolyID, PolynomialType,
+};
+use powdr_number::FieldElement;
+
+/// Accessor for trace values.
+pub trait TraceValues<T> {
+    fn get(&self, poly_id: &AlgebraicReference) -> T;
+}
+
+/// A simple container for trace values.
+pub struct OwnedTraceValues<T> {
+    pub values: BTreeMap<PolyID, Vec<T>>,
+}
+
+/// A view into the trace values for a single row.
+pub struct RowTraceValues<'a, T> {
+    trace: &'a OwnedTraceValues<T>,
+    row: usize,
+}
+
+impl<T> OwnedTraceValues<T> {
+    pub fn new(
+        pil: &Analyzed<T>,
+        witness_columns: Vec<(String, Vec<T>)>,
+        fixed_columns: Vec<(String, Vec<T>)>,
+    ) -> Self {
+        let mut columns_by_name = witness_columns
+            .into_iter()
+            .chain(fixed_columns)
+            .collect::<BTreeMap<_, _>>();
+        let values = pil
+            .committed_polys_in_source_order()
+            .chain(pil.constant_polys_in_source_order())
+            .flat_map(|(symbol, _)| symbol.array_elements())
+            .map(|(name, poly_id)| {
+                let column = columns_by_name
+                    .remove(&name)
+                    .unwrap_or_else(|| panic!("Missing column: {name}"));
+                (poly_id, column)
+            })
+            .collect();
+        Self { values }
+    }
+
+    pub fn row(&self, row: usize) -> RowTraceValues<T> {
+        RowTraceValues { trace: self, row }
+    }
+}
+
+impl<'a, F: FieldElement> TraceValues<F> for RowTraceValues<'a, F> {
+    fn get(&self, column: &AlgebraicReference) -> F {
+        match column.poly_id.ptype {
+            PolynomialType::Committed | PolynomialType::Constant => {
+                let column_values = self.trace.values.get(&column.poly_id).unwrap();
+                let row = (self.row + column.next as usize) % column_values.len();
+                column_values[row]
+            }
+            PolynomialType::Intermediate => unreachable!(
+                "Intermediate polynomials should have been handled by ExpressionEvaluator"
+            ),
+        }
+    }
+}
+
+/// Evaluates an algebraic expression to a value.
+pub struct ExpressionEvaluator<'a, T, SV> {
+    trace_values: SV,
+    intermediate_definitions: &'a BTreeMap<AlgebraicReferenceThin, Expression<T>>,
+    challenges: &'a BTreeMap<u64, T>,
+    /// Maps intermediate reference to their evaluation. Updated throughout the lifetime of the
+    /// ExpressionEvaluator.
+    intermediates_cache: BTreeMap<AlgebraicReferenceThin, T>,
+}
+
+impl<'a, T, TV> ExpressionEvaluator<'a, T, TV>
+where
+    TV: TraceValues<T>,
+    T: FieldElement,
```

Comment:
The only `FieldElement`-specific stuff is in the implementation of `AlgebraicBinaryOperator::Pow`. Otherwise, this could be `T: Add<Output=T> + Mul<Output=T> + ...`.

I think in the future we could use this to replace functions like [this](https://github.com/powdr-labs/powdr/blob/60860b731782e4f0779fa6a9d6662d76a1163e40/plonky3/src/circuit_builder.rs#L230) in the backends.

We could even replace `PartialExpressionEvaluator` if we implement these operations for `AffineResult<AlgebraicVariable<'a>, T>`!